harness/determined/keras/_enqueuer.py

```diff
@@ -0,0 +1,334 @@
+import abc
+import collections
+import multiprocessing
+import multiprocessing.queues
+import queue
+import threading
+from typing import Any, Deque, Dict, Iterator, Optional, Type, Union
+
+import numpy as np
+import tensorflow as tf
+
+Queue = Union[queue.Queue, multiprocessing.Queue]
+Worker = Union[threading.Thread, multiprocessing.Process]
+
+
+class _Sampler:
+    """
+    _Sampler decides the order of batches to read from a Keras Sequence, similar to a PyTorch
+    Sampler object.  In a dataloader, there is one _Sampler in the main thread/process that decides
+    the order of data loading for all the dataloader worker threads/processes.
+
+    _Sampler therefore makes it possible to shuffle indices without concern for synchronous access,
+    as is the case when trying to implement shuffle from inside of a Keras Sequence inside of the
+    real Keras OrderedEnqueuer.
+    """
+
+    def __init__(
+        self,
+        length: int,
+        shard_rank: int,
+        shard_size: int,
+        shuffle: bool,
+        shuffle_seed: int,
+        prior_batches_trained: int,
+    ) -> None:
+        self.indices = list(range(length))
+        self.shard_size = shard_size
+        self.shuffle = shuffle
+
+        # Each shard has a certain offset from which it yields data.  When the dataset length is
+        # not evenly divisible by the shard size, that offset will change every epoch.
+        # Example:
+        #   let length=10, shard_rank=0, and shard_size=3:
+        #   epoch 1: 0, 3, 6, 9
+        #   epoch 2: 2, 5, 8
+        #   epoch 3: 1, 4, 7
+        #   epoch 4: (same as epoch 1)
+        # In this example, the offset in the first three epochs is 0, then 2, then 1.
+        self.offset = shard_rank
+
+        if self.shuffle:
```

Comment:
> What do you mean? We can't save the state but it would be easy to calculate the state?

Due to the prefetching, there are two timelines going on.  The Sampler might think it is a couple batches into epoch 9 while the model thinks it is a couple of batches from the end of epoch 8.  You always decide to save the state of the model on the model's timeline, but there's no guarantee that if you save the state of the RNG in the Sampler right then that you are getting the correct RNG state.  This is the same reason why we need to implement a distributed snapshot for master restart; because the searcher and the workload sequencers are each on different timelines.

However, it is easy to just start the Sampler from epoch 0 every time and call `_end_epoch()` to the exact number of epochs that the Sampler would have gone through, based on the prior batches processed for the initial workload.  At that point, the next index from the Sampler will be the next index that the model needs.  That's what I mean when I say it's easy to calculate the state.  This is identical to what we do in `SkipBatchSampler` in the pytorch code.